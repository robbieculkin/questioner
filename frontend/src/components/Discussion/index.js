import React, { Component } from 'react';
import axios from 'axios';
import { Link } from 'react-router-dom';

import './index.scss';

import QuestionAgent from '../../logic/questionAgent';

import Card from '../Card';

class Discussion extends Component {
  constructor(props) {
    super(props);

    const questionAgent = new QuestionAgent();

    this.state = {
      nextId: 1,
      questionAgent,
      history: [
        {
          id: 0,
          fromUser: false,
          text: 'NO QUESTION GIVEN'
        }
      ],
    };

    this.handleTextEntry = this.handleTextEntry.bind(this);
  }

  componentDidMount() {
    axios.get('http://localhost:5000/api/v0/questions')
      .then(res => {
          const data = res.data;

          this.setState({
            history: [
              {
                id: 0,
                text: data.question,
                fromUser: false
              }
            ]
          })
      });
  }

  handleTextEntry(text) {
    const { history, nextId, questionAgent } = this.state;

    this.setState({
      history: [
        ...history,
        {
          id: nextId,
          text,
          fromUser: true
        },
        {
          id: nextId + 1,
          text: questionAgent.getQuestion(),
          fromUser: false
        }
      ],
      nextId: nextId + 2
    });
  }

  render() {
    const { history } = this.state;

    return (
      <div className='container centered'>
        <div className='title'>
          <h1>Questioner</h1>
        </div>
        <div className='tile'>
          <Card history={history} onTextEntry={this.handleTextEntry} />
        </div>
        <Link to='/report' className='link'>
          <div className='main-button tile'>End Discussion</div>
        </Link>
      </div>
    );
  }
}

export default Discussion;
