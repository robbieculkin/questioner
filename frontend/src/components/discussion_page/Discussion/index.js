import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import axios from 'axios';
import uuidv1 from 'uuid';

import './index.scss';

import Card from '../Card';

import { post_config } from '../../../config/headers';

class Discussion extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sessionId: uuidv1(),
      history: [
        {
          msgId: 0,
          fromUser: false,
          text: 'Loading...'
        }
      ],
    };

    this.handleTextEntry = this.handleTextEntry.bind(this);
  }

  componentDidMount() {
    const { sessionId } = this.state;

    axios.get('http://localhost:5000/api/v0/questions', {
        params: { sessionId }
      })
      .then(res => {
        const data = res.data;

        this.setState({
          history: [
            {
              msgId: uuidv1(),
              text: data.question,
              fromUser: false
            }
          ]
        });
      })
      .catch(error => {
        console.log(error);
        this.setState({
          history: [
            {
              msgId: uuidv1(),
              text: 'Oops, something went wrong! Please try again.',
              fromUser: false
            }
          ]
        })
      });
  }

  handleTextEntry(text) {
    const { sessionId, history } = this.state;
    const updatePayload = {
      session: {
        sessionId,
        discussion: [
          ...history,
          {
            msgId: uuidv1(),
            text,
            fromUser: true
          }
        ]
      }
    }

    axios.post('http://localhost:5000/api/v0/response',
               updatePayload, post_config)
      .then(res => {
        console.log(res);
      })
      .catch(error => {
        console.log(error);
      });

    this.setState({
      history: [
        {
          msgId: uuidv1(),
          text: 'next question here',
          fromUser: false
        }
      ],
    });
  }

  render() {
    const { sessionId, history } = this.state;

    return (
      <div className='container centered'>
        <div className='title'>
          <h1>Questioner</h1>
        </div>
        <div className='tile'>
          <Card history={history} onTextEntry={this.handleTextEntry} />
        </div>
        <Link to={{ pathname: '/report', state: { sessionId } }} className='link'>
          <div className='main-button tile'>End Discussion</div>
        </Link>
      </div>
    );
  }
}

export default Discussion;
