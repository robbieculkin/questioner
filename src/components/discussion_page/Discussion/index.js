import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import axios from 'axios';
import uuidv1 from 'uuid';
import PropTypes from 'prop-types';

import './index.scss';

import Card from '../Card';

import { post_config } from '../../../config/headers';
import { errorState } from '../../../data/errorData';
import { RESPONSE_URI, QUESTIONS_URI } from '../../../config/api';

class Discussion extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sessionId: props.location.state.sessionId,
      selectedPlay: props.location.state.selectedPlay,
      history: [
        {
          msgId: 0,
          fromUser: false,
          text: 'Loading...'
        }
      ],
    };

    this.handleTextEntry = this.handleTextEntry.bind(this);
  }

  componentDidMount() {
    const { sessionId } = this.state;
    this._mounted = true;

    axios.get(QUESTIONS_URI, { params: { sessionId } })
      .then(res => {
        if (!this._mounted) return;
        this.setState({
          history: [
            {
              msgId: uuidv1(),
              text: res.data.question,
              fromUser: false
            }
          ]
        });
      })
      .catch(error => {
        if (!this._mounted) return;
        console.log(error);
        this.setState(errorState);
      });
  }

  componentWillUnmount() {
    this._mounted = false;
  }

  handleTextEntry(text) {
    const { sessionId, selectedPlay, history } = this.state;
    const updatePayload = {
      session: {
        sessionId,
        selectedPlay,
        discussion: [
          ...history,
          {
            msgId: uuidv1(),
            text,
            fromUser: true
          }
        ]
      }
    }

    // POST question and user response
    axios.post(RESPONSE_URI, updatePayload, post_config)
      .then(() => {
        // GET next question
        axios.get(QUESTIONS_URI, { params: { sessionId } })
          .then(res => {
            this.setState({
              history: [
                {
                  msgId: uuidv1(),
                  text: res.data.question,
                  fromUser: false
                }
              ]
            });
          })
          .catch(error => {
            console.log(error);
            this.setState(errorState);
          });
      })
      .catch(error => {
        console.log(error);
        this.setState(errorState);
      });
  }

  render() {
    const { sessionId, history } = this.state;

    return (
      <div className='container centered'>
        <div className='title'>
          <h1 className='big-title'>Questioner</h1>
        </div>
        <div className='tile'>
          <Card history={history} onTextEntry={this.handleTextEntry} />
        </div>
        <Link to={{ pathname: '/report', state: { sessionId } }}
              className='link'>
          <div className='main-button tile'>End Discussion</div>
        </Link>
      </div>
    );
  }
}

Discussion.propTypes = {
  location: PropTypes.shape({
    state: PropTypes.shape({
      sessionId: PropTypes.string,
      selectedPlay: PropTypes.string
    })
  })
};

export default Discussion;
