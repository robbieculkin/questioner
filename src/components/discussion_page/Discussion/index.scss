.discussion {
  &.right-container {
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    padding-right: 90px;
  }

  .link {
    font-size: 36px;
    position: fixed;
    left: 50px;
    bottom: 50px;
  }

  .fixed-title {
    position: fixed;
    left: 50px;
    top: 0;
    &.hidden-fade {
      opacity: 0;
      visibility: hidden;
      transition: all 0.3s ease-in-out;
    }
    &.visible-fade {
      opacity: 1;
      visibility: visible;
      transition: all 0.3s ease-in-out;
    }
    .left-align {
      text-align: left;
      padding-left: 40px;
    }
  }
}