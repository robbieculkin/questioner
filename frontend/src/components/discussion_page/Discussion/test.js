import React from 'react';
import { shallow } from 'enzyme';
import Discussion from '.';

jest.mock('uuid', () => jest.fn(() => '123uuid'));

describe('Discussion', () => {
  it('should render correctly', () => {
    const discussionProps = { location: { state: { selectedPlay: 'Hamlet' }}};
    const component = shallow(<Discussion {...discussionProps} />);
    expect(component).toMatchSnapshot();
  });
});
