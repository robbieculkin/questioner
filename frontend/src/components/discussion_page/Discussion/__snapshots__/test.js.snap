// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discussion should render correctly 1`] = `
<div
  className="container centered"
>
  <div
    className="title"
  >
    <h1>
      Questioner
    </h1>
  </div>
  <div
    className="tile"
  >
    <Card
      history={
        Array [
          Object {
            "fromUser": false,
            "msgId": 0,
            "text": "Loading...",
          },
        ]
      }
      onTextEntry={[Function]}
    />
  </div>
  <Link
    className="link"
    replace={false}
    to={
      Object {
        "pathname": "/report",
        "state": Object {
          "sessionId": "123uuid",
        },
      }
    }
  >
    <div
      className="main-button tile"
    >
      End Discussion
    </div>
  </Link>
</div>
`;
